common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java

```diff
@@ -339,11 +339,15 @@ public static UTF8String lowercaseReplace(final UTF8String src, final UTF8String
    * @return the uppercase string
    */
   public static UTF8String toUpperCase(final UTF8String target) {
-    return UTF8String.fromString(toUpperCase(target.toString()));
+    if (target.isFullAscii()) return target.toUpperCaseAscii();
+    return toUpperCaseSlow(target);
   }
 
-  public static String toUpperCase(final String target) {
-    return UCharacter.toUpperCase(target);
+  private static UTF8String toUpperCaseSlow(final UTF8String target) {
+    // Note: In order to achieve the desired behaviour, we use the ICU UCharacter class to
+    // convert the string to uppercase, which only accepts a Java strings as input.
+    // TODO: All UTF8String -> String conversions should use `makeValid` (SPARK-48715)
```

Comment:
I believe the proper etiquette for `TODO` comments is to either add a JIRA or a user name.
So this should probably be:
```java
// TODO(SPARK-48715): All UTF8String -> String conversions should use `makeValid`
```